// @discloudapp/api-types/payload/v2/DiscloudConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.discloudConfigScopes = exports.APTPackages = exports.APT = void 0;
exports.APT = {
    canvas: [
        "libcairo2",
        "libpango1.0-dev",
        "libjpeg-dev",
        "libgif-dev",
        "librsvg2-dev",
        "libgbm-dev",
    ],
    ffmpeg: ["ffmpeg"],
    java: ["default-jre"],
    libgl: ["libsm6", "libxext6"],
    openssl: ["pkg-config", "libssl-dev"],
    puppeteer: [
        "libasound2",
        "libatk-1.0-0",
        "libatk-bridge2.0-0",
        "libcairo2",
        "libcups2",
        "libdrm-dev",
        "libdrm2",
        "libgbm-dev",
        "libgbm1",
        "libglib2.0-0",
        "libgtk-3-0",
        "libnss3",
        "libpango-1.0-0",
        "libx11-xcb-dev",
        "libxcomposite1",
        "libxdamage1",
        "libxfixes3",
        "libxkbcommon0",
        "libxrandr2",
        "libxshmfence-dev",
    ],
    tools: ["cmake", "curl", "git", "make", "wget"],
};
exports.APTPackages = Object.keys(exports.APT);
exports.discloudConfigScopes = [
    "APT",
    "AUTORESTART",
    "AVATAR",
    "ID",
    "MAIN",
    "NAME",
    "RAM",
    "TYPE",
    "VERSION",
];
