// discloud.app/dist/structures/AppBackup.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_fs_1 = require("node:fs");
const node_process_1 = require("node:process");
const undici_1 = require("undici");
const Base_1 = __importDefault(require("./Base"));
class AppBackup extends Base_1.default {
    /**
     * The backup status
     */
    status;
    /**
     * The backup {@link Buffer}
     */
    data;
    constructor(discloudApp, data) {
        super(discloudApp);
        this._patch(data);
    }
    _patch(data) {
        this.id = data.id;
        this.url = data.url;
        if ("status" in data)
            this.status = data.status;
        return super._patch(data);
    }
    /**
     * Make backup of your application
     *
     * @param path - Backup path
     * @param fileName - Backup file name
     */
    async download(path = (0, node_process_1.cwd)(), fileName = this.id) {
        this.data = await (0, undici_1.fetch)(this.url).then(res => res.arrayBuffer()).then(Buffer.from);
        const file = `${path}/${fileName}.${this.url.split(".").at(-1)}`;
        (0, node_fs_1.writeFileSync)(file, this.data);
        if ((0, node_fs_1.existsSync)(file))
            this.url = file;
        return this;
    }
}
exports.default = AppBackup;
