// discloud.app/dist/structures/AppUploaded.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Base_1 = __importDefault(require("./Base"));
class AppUploaded extends Base_1.default {
    constructor(discloudApp, data) {
        super(discloudApp);
        this._patch(data);
    }
    get app() {
        return this.discloudApp.apps.cache.get(this.appId);
    }
    _patch(data) {
        if ("addedAtTimestamp" in data)
            this.addedAtTimestamp = data.addedAtTimestamp;
        if ("avatarURL" in data)
            this.avatarURL = data.avatarURL;
        if ("autoRestart" in data)
            this.autoRestart = data.autoRestart;
        if ("id" in data)
            this.appId = data.id;
        if ("lang" in data)
            this.lang = data.lang;
        if ("mainFile" in data)
            this.mainFile = data.mainFile;
        if ("name" in data)
            this.name = data.name;
        if ("ram" in data)
            this.ram = data.ram;
        if ("type" in data)
            this.type = data.type;
        if ("version" in data)
            this.version = data.version;
        return super._patch(data);
    }
}
exports.default = AppUploaded;
