// discloud.app/dist/structures/TeamAppStatus.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("@discloudapp/util");
const Base_1 = __importDefault(require("./Base"));
class TeamAppStatus extends Base_1.default {
    constructor(discloudApp, data) {
        super(discloudApp);
        this._patch(data);
    }
    get app() {
        return this.discloudApp.teamApps.cache.get(this.appId);
    }
    fetch() {
        return this.discloudApp.teamApps.status(this.appId);
    }
    _patch(data) {
        if ("container" in data)
            this.container = data.container;
        if ("cpu" in data)
            this.cpu = data.cpu;
        if ("id" in data)
            this.appId = data.id;
        if ("last_restart" in data)
            this.lastRestart = data.last_restart;
        if ("memory" in data) {
            this.memory = data.memory;
            const matched = data.memory.match(/[\d.]+/g) ?? [];
            this.memoryUsage = (0, util_1.calculatePercentage)(matched[0], matched[1]);
        }
        if ("netIO" in data)
            this.netIO = data.netIO;
        if ("ssd" in data)
            this.ssd = data.ssd;
        if ("startedAt" in data) {
            this.startedAt = new Date(data.startedAt);
            this.startedAtTimestamp = this.startedAt.valueOf();
        }
        return this;
    }
}
exports.default = TeamAppStatus;
