// discloud.app/dist/structures/TeamApp.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseTeamApp_1 = __importDefault(require("./BaseTeamApp"));
const TeamAppStatus_1 = __importDefault(require("./TeamAppStatus"));
class TeamApp extends BaseTeamApp_1.default {
    /**
     * Your permissions to manage this application.
     */
    perms = [];
    constructor(discloudApp, data) {
        super(discloudApp, data);
        this._patch(data);
    }
    _patch(data) {
        if ("exitCode" in data)
            this.exitCode = data.exitCode;
        if ("lang" in data)
            this.lang = data.lang;
        if ("name" in data)
            this.name = data.name;
        if ("online" in data)
            this.online = data.online;
        if ("perms" in data)
            this.perms = data.perms;
        if ("ramKilled" in data)
            this.ramKilled = data.ramKilled;
        this.status ??= new TeamAppStatus_1.default(this.discloudApp, data);
        return super._patch(data);
    }
}
exports.default = TeamApp;
