// discloud.app/dist/structures/User.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const v2_1 = require("@discloudapp/api-types/v2");
const __1 = require("..");
const Base_1 = __importDefault(require("./Base"));
class User extends Base_1.default {
    /**
     * Your applications ID
     */
    appIDs = [];
    /**
     * Your custom domains on Discloud
     */
    customdomains = [];
    /**
     * Your subdomains on Discloud
     */
    subdomains = [];
    constructor(discloudApp, data) {
        super(discloudApp);
        this._patch(data);
    }
    _patch(data) {
        if ("apps" in data)
            this.appIDs = data.apps;
        if ("customdomains" in data)
            this.customdomains = data.customdomains;
        if ("userID" in data)
            this.id = data.userID;
        if ("locale" in data)
            this.locale = data.locale;
        if ("plan" in data)
            this.plan = data.plan;
        if ("planDataEnd" in data) {
            this.planDataEnd = new Date(data.planDataEnd);
            this.planDataEndTimestamp = this.planDataEnd.valueOf();
        }
        if ("ramUsedMb" in data)
            this.ramUsedMb = data.ramUsedMb;
        if ("subdomains" in data)
            this.subdomains = data.subdomains;
        if ("totalRamMb" in data)
            this.totalRamMb = data.totalRamMb;
        if ("ramUsedMb" in this && "totalRamMb" in this)
            this.ramUsage = (0, __1.calculatePercentage)(this.ramUsedMb, this.totalRamMb);
        return super._patch(data);
    }
    get apps() {
        return this.discloudApp.apps;
    }
    /**
     * Set your locale on Discloud
     *
     * @param locale - Your locale like "en-US".
     * @see {@link LocaleString}
     * @returns Promise {@link Boolean}
     */
    async setLocale(locale) {
        const data = await this.discloudApp.rest.put(v2_1.Routes.locale(locale));
        if ("locale" in data)
            this.locale = data.locale;
        return data.locale === locale;
    }
    /**
     * Fetch a user from Discloud API
     *
     * @returns Promise {@link User}
     */
    async fetch() {
        const data = await this.discloudApp.rest.get(v2_1.Routes.user());
        return this._patch(data.user);
    }
    /**
     * @returns user id
     */
    toString() {
        return this.id;
    }
}
exports.default = User;
