// @discloudapp/api-types/rest/v2/routes/App.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.App = void 0;
exports.App = new class App {
    /**
     * - GET - `/app/{appID}`
     */
    app(appID) {
        return `/app/${appID}`;
    }
    /**
     * - PUT - `/app/{appID}/apt`
     * - DELETE - `/app/{appID}/apt`
     */
    apt(appID) {
        return `/app/${appID}/apt`;
    }
    /**
     * - GET - `/app/{appID}/backup`
     */
    backup(appID) {
        return `/app/${appID}/backup`;
    }
    /**
     * - PUT - `/app/{appID}/commit`
     */
    commit(appID) {
        return `/app/${appID}/commit`;
    }
    /**
     * - DELETE - `/app/{appID}/delete`
     */
    delete(appID) {
        return `/app/${appID}/delete`;
    }
    /**
     * - GET - `/app/{appID}/logs`
     */
    logs(appID) {
        return `/app/${appID}/logs`;
    }
    /**
     * - PUT - `/app/{appID}/ram`
     */
    ram(appID) {
        return `/app/${appID}/ram`;
    }
    /**
     * - PUT - `/app/{appID}/restart`
     */
    restart(appID) {
        return `/app/${appID}/restart`;
    }
    /**
     * - PUT - `/app/{appID}/start`
     */
    start(appID) {
        return `/app/${appID}/start`;
    }
    /**
     * - GET - `/app/{appID}/status`
     */
    status(appID) {
        return `/app/${appID}/status`;
    }
    /**
     * - PUT - `/app/{appID}/stop`
     */
    stop(appID) {
        return `/app/${appID}/stop`;
    }
    team(appID, modID) {
        return modID ? `/app/${appID}/team/${modID}` : `/app/${appID}/team`;
    }
};
