// @discloudapp/api-types/rest/v2/routes/Team.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Team = void 0;
exports.Team = new class Team {
    /**
     * - GET - `/team`
     */
    team() {
        return "/team";
    }
    /**
     * - GET - `/app/{appID}/backup`
     */
    backup(appID) {
        return `/team/${appID}/backup`;
    }
    /**
     * - PUT - `/app/{appID}/commit`
     */
    commit(appID) {
        return `/team/${appID}/commit`;
    }
    /**
     * - GET - `/team/{appID}/logs`
     */
    logs(appID) {
        return `/team/${appID}/logs`;
    }
    /**
     * - PUT - `/team/{appID}/ram`
     */
    ram(appID) {
        return `/team/${appID}/ram`;
    }
    /**
     * - PUT - `/app/{appID}/restart`
     */
    restart(appID) {
        return `/team/${appID}/restart`;
    }
    /**
     * - PUT - `/app/{appID}/start`
     */
    start(appID) {
        return `/team/${appID}/start`;
    }
    /**
     * - PUT - `/app/{appID}/status`
     */
    status(appID) {
        return `/team/${appID}/status`;
    }
    /**
     * - PUT - `/app/{appID}/stop`
     */
    stop(appID) {
        return `/team/${appID}/stop`;
    }
};
